src/App.tsx

```diff
@@ -1,61 +1,174 @@
+import { useState } from 'react';
 import './App.scss';
 
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+import { TodoList } from './components/TodoList';
+import { User } from './types/User';
+import { Todo } from './types/Todo';
+
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+
+function getUser(userId: number): User | null {
+  return usersFromServer.find((user) => user.id === userId) || null;
+}
+
+export const todos: Todo[] = todosFromServer.map((todo) => ({
+  ...todo,
+  user: getUser(todo.userId),
+}));
+
+const allowedChars
+    = '0123456789abcdefghigklmnopqrstuvwxyz абвгдеёжзийклмнопрстуфхцчшщъыьэюя';
+
+export const App: React.FC = () => {
+  const [todoList, setTodoList] = useState(todos);
+
+  const [newTodo, setNewTodo] = useState({
+    title: '',
+    userName: '',
+    isTitleValid: true,
+    isUserNameValid: true,
+  });
+
+  const handleChange = (
+    e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>,
+  ) => {
+    const { name, value } = e.target;
+
+    if (
+      name === 'title'
+      && allowedChars.includes(value.slice(-1).toLowerCase())
+    ) {
+      setNewTodo((todo) => ({
+        ...todo,
+        [name]: value,
+        isTitleValid: true,
+      }));
+    }
+
+    if (name === 'userName') {
+      setNewTodo((todo) => ({
+        ...todo,
+        [name]: value,
+        isUserNameValid: true,
+      }));
+    }
+  };
+
+  const handleAddTodo = () => {
+    if (!newTodo.title && !newTodo.userName) {
+      setNewTodo((todo) => ({
+        ...todo,
+        isTitleValid: false,
+        isUserNameValid: false,
+      }));
+
+      return;
+    }
+
+    if (!newTodo.title) {
+      setNewTodo((todo) => ({
+        ...todo,
+        isTitleValid: false,
+      }));
+
+      return;
+    }
+
+    if (!newTodo.userName) {
+      setNewTodo((todo) => ({
+        ...todo,
+        isUserNameValid: false,
+      }));
+
+      return;
+    }
+
+    const lastTodoId = Math.max(...todoList.map((todo) => todo.id));
+
+    const todoUser: User | undefined = usersFromServer.find(
+      (user) => user.name === newTodo.userName,
+    );
+
+    const todoToAdd: Todo = {
+      id: lastTodoId + 1,
+      title: newTodo.title,
+      userId: todoUser?.id,
+      completed: false,
+      user: todoUser,
+    };
+
+    setTodoList((prevTodos) => {
+      prevTodos.push(todoToAdd);
+
+      return [...prevTodos];
+    });
+
+    setNewTodo({
+      title: '',
+      userName: '',
+      isTitleValid: true,
+      isUserNameValid: true,
+    });
+  };
 
-export const App = () => {
   return (
     <div className="App">
       <h1>Add todo form</h1>
 
-      <form action="/api/users" method="POST">
+      <form
+        action="/api/users"
+        method="POST"
+        onSubmit={(e) => {
```

Comment:
it isn't correct form flow,
```suggestion
        onSubmit={handleAddTodo}
```

Reply:
and add preventDefault inside handleAddTodo